In [1]:
import pandas as pd
import numpy as np

In [3]:
wine= pd.read_csv("D:\Datascience\winequality-red.csv")

In [4]:
wine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [6]:
wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [7]:
wine.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [8]:
x=wine.drop(columns=['quality'])

In [9]:
y=wine[['quality']]

In [10]:
from sklearn.model_selection import train_test_split

In [12]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=42)

In [13]:
len(x_train),len(x_test),len(y_train),len(y_test)

(1119, 480, 1119, 480)

In [14]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [15]:
dt = DecisionTreeClassifier()
model_dt = dt.fit(x_train,y_train)

In [16]:
rf = RandomForestClassifier(n_estimators=200,random_state=42)
model_rf = rf.fit(x_train,y_train)

<ipython-input-16-e9ee33ec8ee4>:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  model_rf = rf.fit(x_train,y_train)


In [21]:
y_test

quality  Predict_DT  Predict_RF
803         6           5           5
124         5           4           5
350         6           5           5
682         5           6           5
1326        6           6           6
...       ...         ...         ...
1468        7           7           7
495         8           8           8
1325        6           6           6
514         7           6           6
576         4           5           5

[480 rows x 3 columns]

In [17]:
y_test['Predict_DT'] = model_dt.predict(x_test)

<ipython-input-17-e1bfd4eb0037>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_test['Predict_DT'] = model_dt.predict(x_test)


In [18]:
y_test['Predict_RF'] = model_rf.predict(x_test)

<ipython-input-18-efc987ea5d9c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_test['Predict_RF'] = model_rf.predict(x_test)


In [19]:
from sklearn.metrics import classification_report

In [22]:
print(classification_report(y_test['quality'],y_test['Predict_DT']))


              precision    recall  f1-score   support

           3       0.00      0.00      0.00         1
           4       0.06      0.06      0.06        17
           5       0.64      0.63      0.63       195
           6       0.56      0.58      0.57       200
           7       0.46      0.44      0.45        61
           8       0.67      0.33      0.44         6

    accuracy                           0.56       480
   macro avg       0.40      0.34      0.36       480
weighted avg       0.56      0.56      0.56       480



In [23]:
print(classification_report(y_test['quality'],y_test['Predict_RF']))


              precision    recall  f1-score   support

           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00        17
           5       0.73      0.76      0.75       195
           6       0.63      0.70      0.66       200
           7       0.57      0.48      0.52        61
           8       0.50      0.17      0.25         6

    accuracy                           0.66       480
   macro avg       0.40      0.35      0.36       480
weighted avg       0.64      0.66      0.65       480



C:\Users\Mrunmayee\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## we can conclude that random forest is more suitable for this case than decision tree. As accuracy for random forest is 66% which is more than decision tree accuracy which is 56%